### ISSUE TITLE
Incorrect Date format on Import

### ISSUE THREAD

**Issue by gmulhall** [moneymanagerex/moneymanagerex#4317]
MMEX version:
 - [x] 1.5.14 64 bit
 - [ ] 1.6.x

Note: bug reporters are expected to have verified the bug still exists
either in the last stable version of MMEX or on updated development code
(master branch).

Operating System:
 - [x] Windows

# Description of the bug

When open the import screen to import a QIF file I select the file and the account to import to. The date format is already showing and in my case that is DD/MM/YYYY so I do not select it again from the drop down.
However frequently but not always the program thinks the data is in MM/DD/YYYY format.
When this happens some data is rejected (if the month is > 12) and some data is then imported with the wrong data (eg. 1st of April dates become 4th of January for example) .

Add screenshot or other files if needed.

# Reproduction

Is the bug reproducible? 
 - [ ] Always 
 - [X] Randomly 
 - [ ] Happened only once

Reproduction steps:

1.  From the menue go to File -> Import -> QIF File
2.  Select the file set the account - leave the date format at whatever is showing - but ensure this was correctly set to what is required from your last import 
3.  Import you file and check if all the data imports OK

…

**Comment by vomikan** [comment]
i think it's not a bug. It depend on current date and dates ib the input file.
for example in the file 03/04/2022 it may be 03 april or 04 march.
The more dates there are in the incoming file, the more likely it is to guess which date mask will fit better.

**Comment by gmulhall** [comment]
Hi,
At the bottom is the .qif file I attempted to load - and please rember that the import dialogue was showing DD/MM/YYY as the date format remembered from the last use. 
Why would MMEx switch to MM/DD/YYYY format of it's own accord when it's being told to use DD/MM/YYYY ? And if it did switch and then reject data why didn't it give a message to the effect ?

In my case the records with valid DD/MM/YYY below dates were all rejected and I then had to manually re-enter them
D19/04/2022
D14/04/2022
D13/04/2022

All the other records had to be manually located and corrected because the day/month had been switched. 
D6/04/2022
D8/04/2022
D11/04/2022
D11/04/2022
D11/04/2022
D1/05/2022

All of this because the program did not use the date format specified. It would have been far easier for me to enter all the data manually. 

I'm a longtime user (and lover) of MMEx but any new user of MMEx would look at what happened to them probably given up on MMEx as being more trouble than it is worth if it can't import data reliably.

If this is not going to be rectified then could you at least explain what I do to avoid this please ?  
Do I have to always select the date format each time before the import? OR is there a problem with a date given as 1/05/2022 not 01/05/2022 ? ie Leading zeros must be given even when zero.

Thanks in anticipation,  

_________________________________________________________________________________
!Type:Bank
D1/05/2022
U-5
T-5
PCBA
LBank Charge
MAccount Fee
^
D19/04/2022
U135
T135
PRCTrader
LPersonal:Geoff
MSale of CF Racewing
^
D14/04/2022
U-385
T-385
PConnect Plumbing Services
LHousehold:Garden
MGarden Tap Repair
^
D13/04/2022
U-42.5
T-42.5
PAct Pharmacies
LMedical:Chemist
MBotox
^
D11/04/2022
U67.8
T67.8
PMedicare
LMedical
MDirect Credit 002221 MCARE BENEFITS 288057472 CYWQ
^
D11/04/2022
U-74.94
T-74.94
PEnergy Locals
LHousehold:Electricity
MDirect Debit 507426 ENERGY LOCALS 480816
^
D11/04/2022
U-192
T-192
PSouthEast Window Cleaning
LHousehold:Garden
MPressure washing of concrete pavers
^
D8/04/2022
U77.65
T77.65
PMedicare
LMedical
MDirect Credit 002221 MCARE BENEFITS 287050555 CYWQ
^
D6/04/2022
U27.7
T27.7
PMedicare
LMedical
MDirect Credit 002221 MCARE BENEFITS 285063645 AABK

**Comment by tactilis** [comment]
In February, there was this 'wrong date format on CSV import bug' that looks similar:

[https://forum.moneymanagerex.org/viewtopic.php?t=10555](https://forum.moneymanagerex.org/viewtopic.php?t=10555)

I was going to write a step by step guide to reproducing it but Nikolay picked it up anyway and fixed it quickly:

[https://github.com/moneymanagerex/moneymanagerex/issues/4171](https://github.com/moneymanagerex/moneymanagerex/issues/4171)

I'm wondering whether the QIF date format problem that @gmulhall is seeing is related?

**Comment by vomikan** [comment]
When trying to import the provided file, the date mask is immediately suggested DD/MM/YYYY. It switches from the default value of DD/MM/YYYY when the QIF file is selected.

![image](https://user-images.githubusercontent.com/6836805/167095551-c471dfe0-3b3b-4ca4-931f-7f7604a4d3df.png)

Other words, I can't reproduce this issue.

**Comment by gmulhall** [comment]
Hi,

In my version (Zip file not installer) the date does not revert to MM/DD/YYYY when the file is selected. See attached screen print.

![2022-05-11_110845](https://user-images.githubusercontent.com/64715676/167749714-a1a41c86-80e2-45ff-b16e-9816b824a35b.jpg)

If it did switch I would regard that as a bug. The Wikipedia spec for Quicken Interchange format gives no indication that QIF files must be month day year. 

Geoff

**Comment by gmulhall** [comment]
Just re-reading vomikan's post

"When trying to import the provided file, the date mask is immediately suggested DD/MM/YYYY. It switches from the default value of DD/MM/YYYY when the QIF file is selected."

This seems to be saying that the date format switches from DD/MM/YYYY to itself. That isn't then a switch then ???

My observation is that DD/MM/YYYY shows on opening the dialogue (presumably left from my past use of MMEx) - so it appears to me that I don't need to select it again - it's already set to what I want. The date format on the actual import in my case was MM/DD/YYYY causing me to have to re-enter some data and fix other data BUT the dialogue after the import still shows DD/MM/YYYY as you can see from my screen print.

What do I need to do to ensure DD/MM/YYYY is honoured ?

**Comment by vomikan** [comment]
Sorry, it's my mistake. The default value is "YYYY-MM-DD"

**Comment by vomikan** [comment]
It should be noted that if the drop-down menu with masks is left untouched, then the most appropriate value is selected automatically when opening the file.
If you select any value in the dropdown menu "Date Format", it will be used regardless of the date format in the file.

**Comment by gmulhall** [comment]
From Vomikan above
"It should be noted that if the drop-down menu with masks is left untouched, then the most appropriate value is selected automatically when opening the file.
If you select any value in the dropdown menu "Date Format", it will be used regardless of the date format in the file.

P.S. Other words, please, do not toche the "Date Format" dropdown menu when the QIF Dialog open."

In other words I must select the date format every time I open the import dialogue _even though it is showing the format I want_ and I must not change the date format after the qif file is selected.

Attached below is the file for testing. Note the file was a .qif file but I've renamed it as a .txt file else Github does not let me upload it.
 
[CBA Out May 2022.txt](https://github.com/moneymanagerex/moneymanagerex/files/8673994/CBA.Out.May.2022.txt)

**Comment by vomikan** [comment]
1)
![image](https://user-images.githubusercontent.com/6836805/168019919-2a96f589-617c-4790-ac69-2325ca55a164.png)

2)
![image](https://user-images.githubusercontent.com/6836805/168020246-066e193e-e38a-4601-b10c-d2fa99a1523f.png)

No matter how hard I try, I can't find an error in the code.

**Comment by gmulhall** [comment]
Thanks for looking at this. I took the file I uploaded and cut it to just 5 transations and with some that would reject if the format was switching to MM/DD/YYYY
[Test.qif.txt](https://github.com/moneymanagerex/moneymanagerex/files/8683224/Test.qif.txt)
The date format showing was DD/MM/YYYY and I did not select it.
This file loaded OK.
Todays date is 13 May. Does MMEx check the system date to derive the format. For the first 12 days of the month the format is ambiguous but not after the 12 day of the month?
Thanks again.
Geoff

**Comment by vomikan** [comment]
The following algorithm is used to determine the date mask. There is a set of date masks. When reading a QIF file, for preview, the date (as text) is extracted and converted to a date (as a number) using each mask. If the conversion passed without an error, then such a mask continues to participate in the trial conversion of dates. If, as a result of date conversion, it is obtained in the future, then the rating of this mask is lower than that of the one in the past. This is done in order to guess which mask fits more DD/MM/YYYY or MM/DD/YYYY. Therefore, the system date matters. And the shorter the source file, the less likely it is to determine the date correctly. If the dates refer to the beginning of the month, then the probability also decreases.

P.S. Thanks to your perseverance, I found an error in the algorithm.

**Comment by vomikan** [comment]
@gmulhall , please try the program with corrections.

https://ci.appveyor.com/project/moneymanagerex/moneymanagerex/builds/43532297/job/5un4kkv9r0neet27/artifacts

**Comment by vomikan** [comment]
Here is an example file that is imported today with the mask MM/DD/YYYY
But on the first of May, the same file would be imported with the mask DD/MM/YYYY
```
!Type:Bank
D4/04/2022
U1.0
T1.0
PMedicare
LMedical
MDirect Credit 002221 MCARE BENEFITS 285063645 AABK
^
D5/04/2022
U1.0
T1.0
PMedicare
LMedical
MDirect Credit 002221 MCARE BENEFITS 287050555 CYWQ
^
```

**Comment by gmulhall** [comment]
Thanks for looking at this in such detail.
I'll try the fix version ASAP I have some time - likely tomorrow.
Thanks again,
Geoff

**Comment by gmulhall** [comment]
Hi,
I tested this today importing 55 transactions and everything worked as perfectly. The file was all DD/MM/YYYY format dates. The new dialogue which has the browse button instead of Select File is an nice I think.
At the end of the import the message showing which file format had been used was nice too.
I then repeated the test using the same import file but selected MM/DD/YYYY and 26 of the transactions were rejected and the program gave a list of the rejections - so that was nice too.
 
Thanks for all your work on this. Really Appreciated. I've made a MMEx donation via PayPal :)

**Comment by gmulhall** [comment]
Hi,
I did one more test where I took a DDMMYYY file and put in invalid dates. The date format showing was DD/MM/YYYY. After the import the date format was switched to MM/DD/YYYY and the change noted in the messages (Nice!). I then switched this back to DD/MM/YYYY and ran the import. All data was rejected. So that was certainly my expected result. File attached below.
[Test.qif.txt](https://github.com/moneymanagerex/moneymanagerex/files/8696609/Test.qif.txt)
Thanks again.
